src/core/api.c

```diff
@@ -1578,6 +1578,94 @@ MsQuicGetParam(
     return Status;
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QUIC_API
+MsQuicHandleAddRef(
+    _In_ _Pre_defensive_
+        HQUIC Handle
+    )
+{
+    CXPLAT_PASSIVE_CODE();
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_ADDREF,
+        Handle);
+
+    if (IS_CONN_HANDLE(Handle)) {
+        #pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+        QUIC_CONNECTION* Connection = (QUIC_CONNECTION*)Handle;
+
+        QUIC_CONN_VERIFY(Connection, !Connection->State.HandleClosed);
+        QUIC_CONN_VERIFY(Connection, !Connection->State.Freed);
+
+        InterlockedIncrement((volatile long*)&Connection->ExternalRefCount);
+    } else if (IS_STREAM_HANDLE(Handle)) {
+        #pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+        QUIC_STREAM* Stream = (QUIC_STREAM*)Handle;
+
+        CXPLAT_TEL_ASSERT(!Stream->Flags.HandleClosed);
+        CXPLAT_TEL_ASSERT(!Stream->Flags.Freed);
+
+        InterlockedIncrement((volatile long*)&Stream->ExternalRefCount);
+    }
```

Comment:
I think we should have a debug assert here so that callers don't use this API for handles that aren't support
```suggestion
    } else {
        CXPLAT_DBG_ASSERTMSG(FALSE, "Only supported for Connection and Stream handles!");
    }   
```